### PR TITLE
add User model, dao, service, AuthenticationService.

### PR DESCRIPTION
**Pull request by Alkey**

### CODE REVIEW

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            transaction = session.beginTransaction();
+            session.save(user);
+            transaction.commit();
+            return user;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't add user " + user + " to DB", e);
+        }
```
Comment (TimoxaV) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java:34]: You should close session in finally block.

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Override
+    public User findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<User> userQuery =
+                    session.createQuery("from User where email = :email", User.class);
+            userQuery.setParameter("email", email);
+            return userQuery.uniqueResult();
```
Comment (TimoxaV) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: checklist says this method should return Optional

File: src/main/java/com/dev/cinema/exception/AuthenticationException.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.dev.cinema.exception;
+
+public class AuthenticationException extends RuntimeException {
```
Comment (TimoxaV) [inline src/main/java/com/dev/cinema/exception/AuthenticationException.java]: checklist - AuthenticationException shouldn’t be RuntimeException

File: src/main/java/com/dev/cinema/model/User.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "user")
```
Comment (TimoxaV) [inline src/main/java/com/dev/cinema/model/User.java]: "user" is a keyword in sql - it's better to name table "users"

**Comment by RomanDubovskyi** [comment]
fix error from checklist before review

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+@Service
+public class UserServiceImpl implements UserService {
+    @Inject
+    private UserDao userDao;
+
+    @Override
+    public User add(User user) {
+        byte[] salt = HashUtil.getSalt();
+        user.setSalt(salt);
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]:
```suggestion
        user.setSalt(HashUtil.getSalt());
```

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        }
+        throw new AuthenticationException("Incorrect password");
+    }
+
+    @Override
+    public User register(String email, String password) throws AuthenticationException {
+        Optional<User> userFromDb = userService.findByEmail(email);
+        if (userFromDb.isPresent()) {
+            throw new AuthenticationException("This email " + email + " is already used");
+        }
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: let's remove this check, we will validate input later on in controller + you throw AuthenticationException while registration - these are different procedures

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import com.dev.cinema.util.HashUtil;
+import java.util.Optional;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java:17]: can we simplify this method, make it with one if clause?

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        Optional<User> userFromDb = userService.findByEmail(email);
+        if (userFromDb.isPresent()
+                && isValid(userFromDb.get().getPassword(), password, userFromDb.get().getSalt())) {
+            return userFromDb.get();
+        }
+        throw new AuthenticationException("Incorrect email: " + email + " or password");
+    }
+
+    @Override
+    public User register(String email, String password) throws AuthenticationException {
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: do we need exception in method signature?

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**